File: src/main/scala/com/github/kayjamlang/core/expressions/Expression.scala
package com.github.kayjamlang.core.expressions

import com.github.kayjamlang.core.opcodes.AccessType

import scala.collection.mutable

abstract class Expression (val accessType: AccessType, val line: Int) extends Cloneable {
  var data = new mutable.HashMap[String, Any]

  @SuppressWarnings(Array("unchecked"))
  @throws[CloneNotSupportedException]
  override protected def clone: Expression = {
    val expression = super.clone.asInstanceOf[Expression]
    expression.data = data clone()
    expression
  }
}

File: src/main/java/com/github/kayjamlang/core/expressions/Expression.java
package com.github.kayjamlang.core.expressions;

import com.github.kayjamlang.core.Stmt;
import com.github.kayjamlang.core.opcodes.AccessType;

public abstract class Expression extends Stmt implements Cloneable {
    /**
     * Type of access
     */
    @Deprecated
    public final AccessType accessType;

    /**
     * Creates Expression
     * @param accessType Type of access
     * @param line Expression line
     */
    protected Expression(AccessType accessType, int line) {
        super(line);
        this.accessType = accessType;
    }

    /**
     * Creates Expression
     * @param line Expression line
     */
    protected Expression(int line) {
        super(line);
        this.accessType = AccessType.NONE;
    }
}